### PR TITLE
nixos/{ollama,open-webui}: minor service update

### PR DESCRIPTION
**Pull request by drupol**
Follow up of https://github.com/NixOS/nixpkgs/pull/316248

## Things done

- Built on platform(s)
  - [x] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- For non-Linux: Is sandboxing enabled in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
  - [ ] `sandbox = relaxed`
  - [ ] `sandbox = true`
- [x] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://github.com/NixOS/nixpkgs/blob/master/pkgs/README.md#package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://nixos.org/manual/nixpkgs/unstable/#ssec-nixos-tests-linking) to the relevant packages
- [x] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [x] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [24.11 Release Notes](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2411.section.md) (or backporting [23.11](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2311.section.md) and [24.05](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2405.section.md) Release notes)
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [x] (Module addition) Added a release notes entry if adding a new NixOS module
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

---

Add a :+1: [reaction] to [pull requests you find important].

[reaction]: https://github.blog/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/
[pull requests you find important]: https://github.com/NixOS/nixpkgs/pulls?q=is%3Aopen+sort%3Areactions-%2B1-desc

### CODE REVIEW

**Comment by drupol** [comment]
@ofborg build ollama open-webui

File: nixos/modules/services/misc/open-webui.nix [hunk 1]
```diff
@@ -12,55 +12,70 @@ in
...
-        description = "Host of open-webui";
+        default = "127.0.0.1";
+        example = "0.0.0.0";
+        description = ''
+          The host address which the Open-WebUI server HTTP interface listens to.
+        '';
       };
 
       port = lib.mkOption {
         type = types.port;
```
Comment (abysssol) [inline nixos/modules/services/misc/open-webui.nix:35]: Is `types.port` a real type? I had to use `types.ints.u16` in ollama's module since I couldn't find a port type.

Comment (drupol) [reply nixos/modules/services/misc/open-webui.nix:35]: Yes, it's used in many places in `nixos/modules/services` :)

File: nixos/modules/services/misc/ollama.nix [hunk 2]
```diff
@@ -116,6 +116,13 @@ in
           Since `ollama run` is mostly a shell around the ollama server, this is usually sufficient.
         '';
       };
+      openFirewall = lib.mkOption {
+        type = types.bool;
+        default = false;
+        description = ''
+          Whether to open the default ports in the firewall for the Ollama server.
+        '';
```
[the remark this answers is not in the document]
Comment (drupol) [reply nixos/modules/services/misc/ollama.nix]: Updated, thanks!

File: nixos/modules/services/misc/open-webui.nix (hunk 1 above)
Comment (abysssol) [reply nixos/modules/services/misc/open-webui.nix:35]: Do you think it's worth switching `types.ints.u16` to `types.port` in ollama's module?

Comment (drupol) [reply nixos/modules/services/misc/open-webui.nix:35]: Yes, just did it in this PR!

**Comment by drupol** [comment]
@ofborg build ollama open-webui

File: nixos/modules/services/misc/ollama.nix [hunk 3]
```diff
@@ -78,11 +78,11 @@ in
         '';
       };
       port = lib.mkOption {
-        type = types.nullOr types.ints.u16;
+        type = types.port;
```
Comment (abysssol) [inline nixos/modules/services/misc/ollama.nix:81]: `services.ollama.port` was intentionally made nullable so `OLLAMA_HOST` could be set to just a host without a port, in case that's ever useful to someone. Why was this removed?

Also, this should be changed as well if you want to completely remove the code for null port:
``` diff
# line 135
environment = cfg.environmentVariables // {
   HOME = cfg.home;
   OLLAMA_MODELS = cfg.models;
-  OLLAMA_HOST =
-    if cfg.port == null then
-      cfg.host
-    else
-      "${cfg.host}:${toString cfg.port}";
+  OLLAMA_HOST = "${cfg.host}:${toString cfg.port}";
```

Comment (drupol) [reply nixos/modules/services/misc/ollama.nix:81]: Job done!

**Comment by drupol** [comment]
@ofborg build ollama open-webui

File: nixos/modules/services/misc/ollama.nix (hunk 3 above)
Comment (abysssol) [reply nixos/modules/services/misc/ollama.nix:81]: but you didn't answer why nullability was removed

Comment (drupol) [reply nixos/modules/services/misc/ollama.nix:81]: Oh right, for two reasons:
1. Because IMHO, it doesn't make sense to me
2. There are very few occurrences of `types.nullOr types.port` in `nixos/services`, compared to `types.port`.

Comment (abysssol) [reply nixos/modules/services/misc/ollama.nix:81]: you think it's unlikely that _anyone_ would _ever_ want to set `OLLAMA_HOST="127.0.0.1"`, without specifying a port?

Comment (drupol) [reply nixos/modules/services/misc/ollama.nix:81]: I believe yes. In that case, there's a default port which is set, so, this will never happen.

Comment (drupol) [reply nixos/modules/services/misc/ollama.nix:81]: If you prefer, I can revert that change, you're maintaining it, I don't want to hijack anything.

Comment (abysssol) [reply nixos/modules/services/misc/ollama.nix:81]: nah, it's fine, I just made it like that when I split `listenAddress` because I was worried that _someone_ would eventually have some need to set just the host without the port, which it didn't support at first until I thought of making `port` nullable.

If you genuinely think it's unnecessary, I honestly would rather remove the added complexity it introduces.

Comment (drupol) [reply nixos/modules/services/misc/ollama.nix:81]: All good then :)

**Review by abysssol — APPROVED** [review]

**Review by shivaraj-bh — APPROVED** [review]
looks good

**Comment by drupol** [comment]
Thank you both!

**Comment by abysssol** [comment]
glad to help how I can, and thank _you_ for improving ollama and open-webui
